### PR TITLE
NIPoPoW Verification

### PR DESCRIPTION
**Pull request by gtklocker**

### CODE REVIEW

File: src/main/scala/org/ergoplatform/local/PoPoWVerifier.scala [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+    def bestChain(): Seq[Header] = {
+        bestProof.map(_.headersChain).getOrElse(Seq())
+    }
+
+    def process(newProof: PoPowProof) {
+        if (newProof.headersChain.head.id == genesisId) {
+            // TODO: order matters here and it shouldn't
+            // only isBetterThan's argument is checked for validity
+            if (!bestProof.map(_.isBetterThan(newProof)).getOrElse(false)) {
```
Comment (kushti) [inline src/main/scala/org/ergoplatform/local/PoPoWVerifier.scala]: So if no bestProof is set, non-valid one could be accepted?

File: src/main/scala/org/ergoplatform/modifiers/history/popow/PoPowProof.scala [hunk 2]
```diff
@@ -44,12 +44,34 @@ case class PoPowProof(m: Int,
...
-    val (thisDivergingChain, thatDivergingChain) = lowestCommonAncestor(headersChain, that.headersChain)
-      .map(h => headersChain.filter(_.height > h.height) -> that.headersChain.filter(_.height > h.height))
-      .getOrElse(headersChain -> that.headersChain)
-    bestArg(thisDivergingChain)(m) > bestArg(thatDivergingChain)(m)
+    if (that.isValid()) {
+      lowestCommonAncestor(headersChain, that.headersChain)
+        .map(h => headersChain.filter(_.height > h.height) -> that.headersChain.filter(_.height > h.height))
+        .map({ case (thisDivergingChain, thatDivergingChain) =>
+          bestArg(thisDivergingChain)(m) > bestArg(thatDivergingChain)(m) })
+        .getOrElse(false)
```
Comment (kushti) [inline src/main/scala/org/ergoplatform/modifiers/history/popow/PoPowProof.scala]: IDEA is proposing to replace map + getOrElse with exists

**Comment by kushti** [comment]
@gtklocker looks good, but please address the comments. Also please comment the code, e.g. ScalaDoc for methods, descriptions of non-trivial steps.

File: src/main/scala/org/ergoplatform/local/PoPoWVerifier.scala (hunk 1 above)
Comment (gtklocker) [reply src/main/scala/org/ergoplatform/local/PoPoWVerifier.scala]: If no bestProof is set, any proof that has a correct genesis is accepted is what this code is saying.

**Pull request merged**